pkg/catalog/catalog.go

```diff
@@ -77,25 +92,16 @@ func (c *ContextCatalog) Upsert(plugin *cli.PluginInfo) error {
 	// When inserting the "global" or "k8s" target we should remove any similar plugin using
 	// the "unknown" target and vice versa.
 	if plugin.Target == configtypes.TargetGlobal || plugin.Target == configtypes.TargetK8s {
-		c.deleteOldTargetEntries(PluginNameTarget(plugin.Name, configtypes.TargetUnknown))
+		delete(c.plugins, PluginNameTarget(plugin.Name, configtypes.TargetUnknown))
 	} else if plugin.Target == configtypes.TargetUnknown {
 		// This could be a global plugin or a k8s plugin (but not both), so let's
 		// delete either pre-existing entry from the catalog
-		c.deleteOldTargetEntries(PluginNameTarget(plugin.Name, configtypes.TargetGlobal))
-		c.deleteOldTargetEntries(PluginNameTarget(plugin.Name, configtypes.TargetK8s))
+		delete(c.plugins, PluginNameTarget(plugin.Name, configtypes.TargetUnknown))
+		delete(c.plugins, PluginNameTarget(plugin.Name, configtypes.TargetUnknown))
```

Comment:
I noticed that we were using `deleteOldTargetEntries` to delete the old target entries.
```
func (c *ContextCatalog) deleteOldTargetEntries(key string) {
	if oldInstallationPath, exists := c.plugins[key]; exists {
		delete(c.sharedCatalog.IndexByPath, oldInstallationPath)
		delete(c.plugins, key)
		delete(c.sharedCatalog.IndexByName, key)
	}
}
```

Now, this function deletes the entry from the shared catalog that it operates on with `delete(c.plugins, key)`.
However, deleting the entries from `IndexByPath` and `IndexByName` should not be done during this step.
```
    delete(c.sharedCatalog.IndexByPath, oldInstallationPath)
    delete(c.sharedCatalog.IndexByName, key)
```
Because we are trying to delete the plugin entry from a specific context catalog and it is possible that the old shared catalog (meaning other existing contexts)  still refers to the pluginInstallionPath that we are trying to delete here.
So, we cannot delete the entry from `IndexByPath` and `IndexByName` at this step. If needed, we should run periodic checks and cleanup unused entries from the catalog cache under `IndexByPath` and `IndexByName`.

Reply:
Ah yes, good catch!  I understand now.  Thanks for the explanation.